lib/ecl/fortio.c

```diff
@@ -451,41 +432,24 @@ bool fortio_is_fortio_file(fortio_type * fortio) {
 */
 
 int fortio_init_read(fortio_type *fortio) {
-  int elm_read;
-  int record_size;
-
-  elm_read = fread(&record_size , sizeof(record_size) , 1 , fortio->stream);
-  if (elm_read == 1) {
-    if (fortio->endian_flip_header)
-      util_endian_flip_vector(&record_size , sizeof record_size , 1);
-
-    return record_size;
-  } else
-    return -1;
+    int32_t record_size;
+    int err = eclfio_sizeof( fortio->stream, fortio->opts, &record_size );
+    return err ? -1 : record_size;
 }
 
 
 bool fortio_data_fskip(fortio_type* fortio, const int element_size, const int element_count, const int block_count) {
-  int headers = block_count * 4;
-  int trailers = block_count * 4;
-  int bytes_to_skip = headers + trailers + (element_size * element_count);
+    return !eclfio_skip( fortio->stream, fortio->opts, block_count );
 
-  return fortio_fseek(fortio, bytes_to_skip, SEEK_CUR);
 }
 
 
 void fortio_data_fseek(fortio_type* fortio, offset_type data_offset, size_t data_element, const int element_size, const int element_count, const int block_size) {
     if(data_element >= element_count) {
         util_abort("%s: Element index is out of range: 0 <= %d < %d \n", __func__, data_element, element_count);
     }
-    {
-      int block_index = data_element / block_size;
-      int headers = (block_index + 1) * 4;
-      int trailers = block_index * 4;
-      offset_type bytes_to_skip = data_offset + headers + trailers + (data_element * element_size);
 
-      fortio_fseek(fortio, bytes_to_skip, SEEK_SET);
-    }
+    eclfio_skip( fortio->stream, fortio->opts, data_element / block_size );
```

Comment:
A chunk of data is generally written in a set of record, where each record has a header and trailer. The blocksizes are quite small: https://github.com/Statoil/libecl/blob/master/lib/ecl/ecl_kw.cpp#L63 - so a large chunk of data is typically split over many blocks. The current master version of the code takes these blocks, and the headers and trailers into account, the current `eclfio_skip()` call is too simple.

Open question: should the low level `eclfio` layer provide functionality to completely hide the fact that data is (typically) split over several blocks? I would vote carefully yes.

Reply:
Hm, so eclipse effectively maintains a maximum size of a record (in the fortran sense) of 4kb, and that longer arrays are just concatenated records?

I would argue that eclfio should in that case not even care, that's it's just a simple I/O layer, and it makes sense for it to be agnostic and tolerant when it comes to these sizes. The fact that records can or will be longer is easily maintained in a higher layer.

Reply:
Eclipse maintains a maximum record size - which is quite small, exactly how small depends on the datatype in question.

Reply:
I would argue that gives us even more reason to make fortio ignorant of it, and let other layers deal with it.

For reading it doesn't matter, because a shorter record and a block would be indistinguishable. Which at that point makes the interface asymmetric (`put` knows about a block size, but `get` does not). There's not too much harm in having another parameter that says block size, I guess, but it does seem a tad clumsy.

I'll think a bit about it, but thanks for the input on the bug, I'll fix it soon.

Reply:
> I would argue that gives us even more reason to make fortio ignorant of it, and let other layers deal with it.

It is not 100% clear to me what you mean here, personally I see conflicting arguments pointing in different directons. I guess it comes to whether this is *fortran* centric functionality/api *eclipse centric*.
 
**record foxus:** Letting `fortio` focus purelly on records makes the api clean and simple to write and maintain. The disadvantage is that as an eclipse centric consumer of this functionality you are not interested in the records - at all; you are interested in IO of full arrays - that is also the semantic level where you are interested in guarantees of atomicity. The records is a detail of eclipse implementation detail which is really not interesting.

A typical array size is O(10000) elements, so in the current implementation that will amount to 100 rounds of setup and teardown in the `put` and `get` methods; I fear that might incur runtime cost in a very inner loop.

2. Letting `fortio` work with full arrays, and also maintain the eclipse expectations of record sizes makes the api/implementation slightly less clean - and as you say potentially asymmetric. But it will be more focused on the actual use case - as I see it.

So - all in all I would vote for this api:
```C
int eclfio_gut(FILE*, const char*, int32 size, void * data);
int eclfio_put(FILE*, const char*, int32 nmemb, int32 record_size, void * data)
```
Where it is assumed that the record functiaonlity is handled fully as an implementation detail inside the `eclfio`implementation.


Finally it should be noted that "random-access" functionality to just read a value here and there is relevant, and becoming more so, the necessary offset calculations require awareness of the blocks.

Reply:
> A typical array size is O(10000) elements, so in the current implementation that will amount to 100 rounds of setup and teardown in the put and get methods; I fear that might incur runtime cost in a very inner loop.

I wouldn't worry about that, as we're about to do IO (which will dominate everything), and it's no more expensive than doing fgetpos/fsetpos, which essentially is just a struct member access. It does add some potential for robustness on writing multiple records in writing an array, because you can retry when it fails on that particular write.

Reply:
Anywho, I think the best approach would be to have one very fortran centric layer, that deals with endianness, and the fstream api directly, and build eclipse-centric functionality on top of that, which now has no need to deal with the C fstream api (at least no so much).

Reply:
Yep, that's one immediate benefit. Maybe fortio will never be useful outside of `ecl`, and I'm honestly ok with that.